### PR TITLE
test async preSerialization with a failed response schema

### LINKED ISSUE #2707: schema validation error on an async response does not trigger the error handler
## 🐛 Bug Report

When using response schema validation on a route with an async handler errors are not caught by the error handler set using `setErrorHandler`, unlike params schema validation.

### Update
This happens only when an async `preSerialization` handler is also present.

## To Reproduce

Let's take a simple working example:
```js
fastify.addHook("preSerialization", async (request, reply, payload) => payload);
fastify.setErrorHandler((error, request, reply) => {
  reply
    .status(error.statusCode || 500)
    .send(error instanceof Error ? { message: error.message, error: error.name } : error);
});
fastify.get(
  '/:id',
  {
    schema: {
      params: {
        type: 'object',
        required: ['id'],
        properties: { id: { type: 'string' } },
      },
      response: {
        200: {
          type: 'object',
          required: ['id'],
          properties: { id: { type: 'string' } },
        },
      },
    },
  },
  async ({ params: { id } }) => ({ id })
);
```

Now what happens if we mess it up?
1. let's replace the route param from `/:id` to `/:blah`
2. let's replace the response from `({ id })` to `({ blah: id })`

This is what produced from both changes:
1. returns the following response, as expected:
```json
{"message":"params should have required property 'id'","error":"Error"}
```
2. returns nothing and doesn't end the request, prints on the server side:
```
(node:1589948) UnhandledPromiseRejectionWarning: Error: "id" is required!
    at Object.$main (eval at build (/home/imoses/workspace/application/node_modules/fast-json-stringify/index.js:158:20), <anonymous>:135:15)
    at serialize (/home/imoses/workspace/application/node_modules/fastify/lib/validation.js:116:41)
    at preserializeHookEnd (/home/imoses/workspace/application/node_modules/fastify/lib/reply.js:349:15)
    at next (/home/imoses/workspace/application/node_modules/fastify/lib/hooks.js:198:7)
    at handleResolve (/home/imoses/workspace/application/node_modules/fastify/lib/hooks.js:209:5)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:1589948) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:1589948) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

## Expected behavior

I expect a validation error on the response schema to be caught and handled by the error handler and not crash the request.

## Your Environment

- *node version*: 14
- *fastify version*: =3.8.0
- *os*: Linux

### PR DESCRIPTION
**Pull request by iMoses**
Fixes #2707 and adds a test for the edge case

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/hooks-async.test.js [hunk 1]
```diff
@@ -649,3 +649,51 @@ test('The this should be the same of the encapsulation level', async t => {
...
+    fastify.inject({
+      method: 'POST',
+      url: '/'
+    }, (err, res) => {
+      t.error(err)
+      t.notEqual(res.statusCode, 200)
+    })
+  })
+
+  t.test('works with async preSerialization', t => {
```
Comment (Eomm) [inline test/hooks-async.test.js:676]: This test dosn't replicate the issue, it return already:

```
{"status":["n","o","t"," ","o","k"]}
200
```

The `required` was working in your examples

Comment (iMoses) [reply test/hooks-async.test.js:676]: but the tests passed, both in the ci and locally O.o
hmm..  i don't mind changing it to whatever you want..
i actually found another bug with `fast-json-stringify` while working on these tests :sweat_smile: 
you can't really use enum in the response schema because the serialize method messes it up

**Review by mcollina — APPROVED** [review]
lgtm

File: test/hooks-async.test.js (hunk 1 above)
Comment (iMoses) [reply test/hooks-async.test.js:676]: while i'm already at it :)

https://github.com/fastify/fast-json-stringify/issues/274

**Review by fox1t — APPROVED** [review]
LGTM

File: lib/hooks.js [hunk 2]
```diff
@@ -195,7 +195,11 @@ function onSendHookRunner (functions, request, reply, payload, cb) {
     }
 
     if (i === functions.length) {
-      cb(null, request, reply, payload)
```
Comment (RafaelGSS) [inline lib/hooks.js:198]: @mcollina this will not decrease the performance since it can't be optimized by v8?

Comment (RafaelGSS) [reply lib/hooks.js:198]: If yes, would be good a benchmark for it, only to be sure how much it will decrease.

Comment (mcollina) [reply lib/hooks.js:198]: Not anymore, try-catch is totally safe since Turbofan.

**Review by RafaelGSS — APPROVED** [review]
LGTM.

File: test/hooks-async.test.js [hunk 3]
```diff
@@ -649,3 +649,51 @@ test('The this should be the same of the encapsulation level', async t => {
...
+      handler (req, reply) { reply.send({ notOk: true }) },
+      schema: { response: { 200: { required: ['ok'], properties: { ok: { type: 'boolean' } } } } }
+    })
+
+    fastify.inject({
+      method: 'POST',
+      url: '/'
+    }, (err, res) => {
+      t.error(err)
+      t.notEqual(res.statusCode, 200)
```
Comment (Eomm) [inline test/hooks-async.test.js:694]: Why has been used `!=200` instead of `===500`?
